presto-spark-base/src/main/java/com/facebook/presto/spark/execution/nativeprocess/HttpNativeExecutionTaskResultFetcher.java

```diff
@@ -136,101 +145,90 @@ public Optional<SerializedPage> pollPage()
+        long bytes = 0;
+        long positionCount = 0;
+        for (SerializedPage page : pages) {
+            if (!isChecksumValid(page)) {
+                throw new PrestoException(
+                        SERIALIZED_PAGE_CHECKSUM_ERROR,
+                        format("Received corrupted serialized page from host %s",
+                                HostAddress.fromUri(workerClient.getLocation())));
             }
-            catch (InterruptedException e) {
-                if (!future.isDone()) {
-                    // This should never happen, but as a sanity check we set exception to future.
-                    client.abortResults();
-                    future.completeExceptionally(e);
-                }
+            bytes += page.getSizeInBytes();
+            positionCount += page.getPositionCount();
+        }
+        log.info("Received %s rows in %s pages from %s", positionCount, pages.size(), workerClient.getTaskUri());
+
+        pageBuffer.addAll(pages);
+        bufferMemoryBytes.addAndGet(bytes);
+        long nextToken = pagesResponse.getNextToken();
+        if (pages.size() > 0) {
+            workerClient.acknowledgeResultsAsync(nextToken);
+        }
+        token = nextToken;
+        if (pagesResponse.isClientComplete()) {
+            workerClient.abortResults();
+            scheduledFuture.cancel(false);
+        }
+        if (!pages.isEmpty()) {
+            synchronized (taskHasResult) {
+                taskHasResult.notifyAll();
             }
-            catch (ExecutionException e) {
-                if (e.getCause() instanceof PageTransportErrorException) {
-                    if (++transportErrorRetries >= MAX_TRANSPORT_ERROR_RETRIES) {
-                        client.abortResults();
-                        future.completeExceptionally(e.getCause());
-                    }
-                }
-                else {
-                    e.printStackTrace();
-                    client.abortResults();
-                    future.completeExceptionally(e);
-                }
+        }
+    }
+
+    private void onFailure(Throwable t)
+    {
+        // record failure
+        try {
+            errorTracker.requestFailed(t);
+        }
+        catch (PrestoException e) {
+            // Entering here means that we are unable to get any results from the CPP process
+            // likely because process has crashed.
+            workerClient.abortResults();
+            stop(false);
+            lastException.set(e);
+            synchronized (taskHasResult) {
+                taskHasResult.notifyAll();
             }
+            return;
+        }
+        ListenableFuture<?> errorRateLimit = errorTracker.acquireRequestPermit();
+        try {
+            // synchronously wait on throttling
+            errorRateLimit.get(maxErrorDuration.toMillis(), TimeUnit.MILLISECONDS);
+        }
+        catch (InterruptedException | ExecutionException | TimeoutException e) {
```

Comment:
Same as other comment. Would be good to expand to all exceptions to avoid silent cancellations of `scheduleAtFixedRate`